noir-projects/noir-protocol-circuits/crates/private-kernel-lib/src/private_call_data_validator.nr

```diff
@@ -0,0 +1,225 @@
+            );
+        }
+    }
+
+    fn validate_contract_address(self) {
+        let contract_address = self.data.call_stack_item.contract_address;
+
+        // TODO(https://github.com/AztecProtocol/aztec-packages/issues/3062): Why is this using a hash function from the stdlib::recursion namespace
+        let private_call_vk_hash = stdlib_recursion_verification_key_compress_native_vk(self.data.vk);
+
+        assert(!contract_address.is_zero(), "contract address cannot be zero");
+        // std::println(f"contract_address={contract_address}");
+        // std::println(f"private_call_vk_hash={private_call_vk_hash}");
+
+        // Recompute the contract class id
+        let computed_private_functions_root = private_functions_root_from_siblings(
+            self.data.call_stack_item.function_data.selector,
+            private_call_vk_hash,
+            self.data.function_leaf_membership_witness.leaf_index,
+            self.data.function_leaf_membership_witness.sibling_path
+        );
+        // std::println(f"computed_private_functions_root={computed_private_functions_root}");
+
+        let computed_contract_class_id = ContractClassId::compute(
+            self.data.contract_class_artifact_hash,
+            computed_private_functions_root,
+            self.data.contract_class_public_bytecode_commitment
+        );
+        // std::println(f"computed_contract_class_id={computed_contract_class_id}");
+
+        // Recompute contract address using the preimage which includes the class_id
+        let computed_partial_address = PartialAddress::compute_from_salted_initialization_hash(
+            computed_contract_class_id,
+            self.data.salted_initialization_hash
+        );
+        // std::println(f"computed_partial_address={computed_partial_address}");
+
+        let computed_address = AztecAddress::compute(self.data.public_keys_hash, computed_partial_address);
+        // std::println(f"computed_address={computed_address}");
+
+        assert(
+            computed_address.eq(contract_address), "computed contract address does not match expected one"
+        );
+    }
+
+    fn validate_call(self) {
+        let call_context = self.data.call_stack_item.public_inputs.call_context;
+
+        let is_own_storage = call_context.storage_contract_address == self.data.call_stack_item.contract_address;
+        if call_context.is_delegate_call {
+            assert(
+                !is_own_storage, "current contract address must not match storage contract address for delegate calls"
+            );
+        } else {
+            assert(is_own_storage, "call stack storage address does not match expected contract address");
+        }
+
+        if call_context.is_static_call {
+            // No state changes are allowed for static calls:
+            assert_eq(self.array_lengths.new_note_hashes, 0, "new_note_hashes must be empty for static calls");
```

Comment:
Check against the lengths gathered in `new()` instead of calling `array_length()` on each array.